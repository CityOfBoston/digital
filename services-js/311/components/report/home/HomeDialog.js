// @flow

import React from 'react';
import { css } from 'glamor';
import { action } from 'mobx';
import { observer } from 'mobx-react';
import Head from 'next/head';

import type { AppStore } from '../../../data/store';

import { MEDIA_LARGE } from '../../style-constants';
import FormDialog from '../../common/FormDialog';
import SectionHeader from '../../common/SectionHeader';
import DescriptionBox from '../../common/DescriptionBox';
import ServiceList from './ServiceList';

export type Props = {
  store: AppStore,
  routeToServiceForm: (code: ?string) => void,
  // eslint-disable-next-line react/no-unused-prop-types
  stage: 'home' | 'service',
};

const DESCRIPTION_HEADER_STYLE = css({
  display: 'none',
  [MEDIA_LARGE]: {
    display: 'block',
  },
});

const NEXT_BUTTON_STYLE = css({
  width: '100%',
  [MEDIA_LARGE]: {
    display: 'none',
  },
});

const SERVICE_PICKER_STYLE = css({
  display: 'none',
  [MEDIA_LARGE]: {
    display: 'block',
  },
});

function renderHome({ store, routeToServiceForm }: Props) {
  return (
    <FormDialog>
      <Head>
        <title>BOS:311 — Report a Problem</title>
      </Head>

      <SectionHeader>311: Boston City Services</SectionHeader>

      <div className="m-v500">
        <div className="g">
          <div className="g--8">
            <h3 className={`stp m-v300 ${DESCRIPTION_HEADER_STYLE.toString()}`}>
              <span className="stp-number">1</span>
              What can we do for you?
            </h3>

            <DescriptionBox
              minHeight={222}
              maxHeight={222}
              text={store.requestForm.description}
              placeholder="How can we help?"
              onInput={action((ev) => { store.requestForm.description = ev.target.value; })}
            />
          </div>

          <div className={`g--44 ${SERVICE_PICKER_STYLE.toString()}`}>
            <h3 className="stp m-v300">Top Service Requests</h3>
            <div style={{ height: 222, overflowY: 'scroll' }}>
              <ServiceList serviceSummaries={store.serviceSummaries} onServiceChosen={routeToServiceForm} />
            </div>
          </div>
        </div>

      </div>

      <button className={`btn ${NEXT_BUTTON_STYLE.toString()}`} onClick={() => { routeToServiceForm(); }}>Next</button>
    </FormDialog>
  );
}

function renderServicePicker({ store, routeToServiceForm }: Props) {
  return (
    <FormDialog>
      <Head>
        <title>BOS:311 — Report a Problem</title>
      </Head>

      <SectionHeader>311: Boston City Services</SectionHeader>

      <div className="m-v500">
        <div className="t--intro">Pick the most related:</div>
        <div style={{ height: 222, overflowY: 'scroll' }}>
          <ServiceList serviceSummaries={store.serviceSummaries} onServiceChosen={routeToServiceForm} />
        </div>
      </div>
    </FormDialog>
  );
}

export default observer(function HomeDialog(props: Props) {
  const { stage } = props;
  switch (stage) {
    case 'home': return renderHome(props);
    case 'service': return renderServicePicker(props);
    default: return null;
  }
});
