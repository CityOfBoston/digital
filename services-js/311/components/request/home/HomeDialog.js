// @flow

import * as React from 'react';
import { action, observable, reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import debounce from 'lodash/debounce';
import type { Context } from 'next';
import Router from 'next/router';

import type { RequestAdditions } from '../../../server/next-handlers';

import type { ServiceSummary } from '../../../data/types';
import type { AppStore, LanguagePreference } from '../../../data/store';
import makeLoopbackGraphql from '../../../data/dao/loopback-graphql';
import type { LoopbackGraphql } from '../../../data/dao/loopback-graphql';
import loadServiceSuggestions from '../../../data/dao/load-service-suggestions';
import loadTopServiceSummaries from '../../../data/dao/load-top-service-summaries';

import FormDialog from '../../common/FormDialog';
import LoadingIcons from '../../common/LoadingIcons';
import { CENTERED_DIALOG_STYLE } from '../../style-constants';
import TranslateDialog from '../translate/TranslateDialog';

import HomePane from './HomePane';
import ChooseServicePane from './ChooseServicePane';

type Stage = 'home' | 'choose';

export type InitialProps = {|
  topServiceSummaries: ServiceSummary[],
  description: string,
  stage: Stage,
  bypassTranslateDialog: boolean,
|};

export type Props = {|
  store: AppStore,
  loopbackGraphql: LoopbackGraphql,
  ...InitialProps,
|};

@observer
export default class HomeDialog extends React.Component<Props> {
  @observable description: string;

  @observable.shallow suggestedServiceSummaries: ?(ServiceSummary[]) = null;
  serviceSuggestionsDisposer: ?Function;

  // Called by RequestLayout
  static async getInitialProps({
    query,
    req,
  }: Context<RequestAdditions>): Promise<InitialProps> {
    const { stage, description, translate } = query;
    const loopbackGraphql = makeLoopbackGraphql(req);

    return {
      topServiceSummaries: await loadTopServiceSummaries(loopbackGraphql, 5),
      stage: stage === 'choose' ? stage : 'home',
      description: description || '',
      bypassTranslateDialog: translate === '0',
    };
  }

  @action
  componentWillMount() {
    this.description = this.props.description;
  }

  componentDidMount() {
    this.serviceSuggestionsDisposer = reaction(
      () => this.description,
      (description: string) => {
        if (description === '') {
          this.suggestedServiceSummaries = [];
        } else {
          this.suggestedServiceSummaries = null;
          this.suggestServices(description);
        }
      },
      {
        fireImmediately: true,
        name: 'update service suggestions from description',
      }
    );

    LoadingIcons.preload();
  }

  componentWillUnmount() {
    if (this.serviceSuggestionsDisposer) {
      this.serviceSuggestionsDisposer();
    }
  }

  suggestServices = debounce(async (description: string) => {
    try {
      const suggestedServiceSummaries = await loadServiceSuggestions(
        this.props.loopbackGraphql,
        description
      );

      runInAction('suggestServices result', () => {
        this.suggestedServiceSummaries = suggestedServiceSummaries;
      });
    } catch (e) {
      runInAction('suggestServices error', () => {
        this.suggestedServiceSummaries = [];
      });
      throw e;
    }
  }, 500);

  @action.bound
  handleDescriptionChanged(ev: SyntheticInputEvent<>) {
    this.description = ev.target.value;
  }

  @action.bound
  routeToChoose() {
    Router.push(
      `/request?stage=choose&description=${encodeURIComponent(
        this.description
      )}`,
      '/request'
    );
  }

  render() {
    const { stage, store, bypassTranslateDialog } = this.props;
    const languages: LanguagePreference[] = store.languages;

    const translateLanguage = TranslateDialog.findLanguage(languages);
    const showTranslate =
      translateLanguage &&
      translateLanguage !== 'en' &&
      !bypassTranslateDialog &&
      stage === 'home';

    const narrow = stage === 'choose';

    if (showTranslate) {
      return <TranslateDialog languages={languages} showContinueInEnglish />;
    }

    return (
      <div className={CENTERED_DIALOG_STYLE}>
        <FormDialog narrow={narrow} noPadding>
          {stage === 'home' && this.renderHome()}
          {stage === 'choose' && this.renderServicePicker()}
        </FormDialog>
      </div>
    );
  }

  renderHome() {
    const { topServiceSummaries, store } = this.props;
    return (
      <HomePane
        store={store}
        description={this.description}
        handleDescriptionChanged={this.handleDescriptionChanged}
        nextFn={this.routeToChoose}
        topServiceSummaries={topServiceSummaries}
      />
    );
  }

  renderServicePicker() {
    const { store: { ui } } = this.props;
    return (
      <ChooseServicePane
        description={this.description}
        suggestedServiceSummaries={this.suggestedServiceSummaries}
        ui={ui}
      />
    );
  }
}
