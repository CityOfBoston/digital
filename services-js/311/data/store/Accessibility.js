// @flow

import { observable, autorun } from 'mobx';

export default class Accessibility {
  el: ?HTMLElement = null;

  @observable message: string = '';
  @observable interrupt: boolean = false;

  messageListenerDisposer: ?Function = null;

  attach() {
    this.el = document.getElementById('ariaLive');

    this.messageListenerDisposer = autorun(
      () => {
        if (this.el) {
          this.el.innerText = this.message;
          this.el.setAttribute(
            'aria-live',
            this.interrupt ? 'assertive' : 'polite'
          );
        }
      },
      {
        name: 'a11y message listener',
      }
    );
  }

  detatch() {
    this.el = null;

    if (this.messageListenerDisposer) {
      this.messageListenerDisposer();
    }
  }
}
