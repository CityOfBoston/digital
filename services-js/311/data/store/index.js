// @flow

import { observable, computed, useStrict } from 'mobx';
import type { IPromiseBasedObservable } from 'mobx-utils';

import type { Service, ServiceSummary, SubmittedRequest } from '../types';
import Question from './Question';

// MobX will enforce that state changes only happen in action blocks.
useStrict(true);

export class ContactInfo {
  @observable required: boolean = false;
  @observable firstName: string = '';
  @observable lastName: string = '';
  @observable email: string = '';
  @observable phone: string = '';

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return !!(this.firstName && this.lastName);
  }
}

export class LocationInfo {
  @observable required: boolean = false;
  @observable address: string = '';
  @observable.shallow location: ?{| lat: number, lng: number |} = null;

  // can be false even when required is false to differentiate between submitting
  // and skipping
  @computed
  get requirementsMet(): boolean {
    return this.address.length > 0;
  }
}

export class AppStore {
  @observable description: string = '';

  @observable contactInfo: ContactInfo = new ContactInfo();
  @observable locationInfo: LocationInfo = new LocationInfo();
  @observable questions: Question[] = [];

  @observable.shallow serviceSummaries: ServiceSummary[];
  serviceCache: Map<string, Service> = observable.shallowMap({});

  apiKeys: {[service: string]: string} = {};
  isPhone: boolean = false;

  @observable requestSubmission: ?IPromiseBasedObservable<SubmittedRequest> = null;

  @observable.ref _currentService: ?Service = null;
  @observable.ref currentServiceError: ?Object = null;

  @computed get currentService(): ?Service {
    return this._currentService;
  }

  set currentService(service: ?Service) {
    if (service === this._currentService) {
      return;
    }

    this._currentService = service;
    this.currentServiceError = null;
    this.questions = [];

    if (service) {
      try {
        this.contactInfo.required = service.contactRequired;
        this.locationInfo.required = service.locationRequired;
        this.questions = Question.buildQuestions(service.attributes);
      } catch (e) {
        this.currentServiceError = e;
      }
    }
  }

  @computed
  get questionRequirementsMet(): boolean {
    return this.questions.filter((q) => !q.required || !q.visible || q.requirementsMet).length === this.questions.length;
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
