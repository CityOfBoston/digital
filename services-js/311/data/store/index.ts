/* global window */

import { observable, configure as mobxConfigure, action } from 'mobx';
import Router from 'next/router';

import RequestSearch from './RequestSearch';
import AddressSearch from './AddressSearch';
import Ui from './Ui';
import BrowserLocation from './BrowserLocation';
import AllServices from './AllServices';
import {
  FetchGraphql,
  ScreenReaderSupport,
  GaSiteAnalytics,
  RouterListener,
} from '@cityofboston/next-client-common';
import { SiteAnalytics } from '@cityofboston/next-client-common/build/SiteAnalytics';

// MobX will enforce that state changes only happen in action blocks.
mobxConfigure({ enforceActions: true });

export type LanguagePreference = {
  code: string;
  region: string | null;
  quality: number;
};

export class AppStore {
  screenReaderSupport: ScreenReaderSupport = new ScreenReaderSupport();
  requestSearch: RequestSearch = new RequestSearch();
  ui: Ui = new Ui();
  browserLocation: BrowserLocation = new BrowserLocation();
  addressSearch: AddressSearch = new AddressSearch();
  allServices: AllServices = new AllServices();
  siteAnalytics: SiteAnalytics = new GaSiteAnalytics();

  languages: LanguagePreference[] = [];

  @observable
  liveAgentAvailable: boolean = (typeof window !== 'undefined' &&
    window.LIVE_AGENT_AVAILABLE) ||
  false;

  _liveAgentButtonId: string = '';

  get liveAgentButtonId(): string {
    return this._liveAgentButtonId;
  }

  set liveAgentButtonId(liveAgentButtonId: string) {
    this._liveAgentButtonId = liveAgentButtonId;

    if (typeof window !== 'undefined') {
      if (window.liveagent) {
        this.listenForLiveAgentEvents(liveAgentButtonId);
      } else {
        if (!window._laq) {
          window._laq = [];
        }

        window._laq.push(() =>
          this.listenForLiveAgentEvents(liveAgentButtonId)
        );
      }
    }
  }

  listenForLiveAgentEvents(liveAgentButtonId: string) {
    window.liveagent!.addButtonEventHandler(
      liveAgentButtonId,
      this.liveAgentEventHandler
    );
  }

  @action.bound
  liveAgentEventHandler(event: string) {
    switch (event) {
      case 'BUTTON_AVAILABLE':
        this.liveAgentAvailable = true;
        break;
      case 'BUTTON_UNAVAILABLE':
        this.liveAgentAvailable = false;
        break;
      default:
        break;
    }
  }
}

let browserStore: AppStore;

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 */
export default function getStore(
  fetchGraphql: FetchGraphql | null = null
): AppStore {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const store = new AppStore();

  if (process.browser) {
    browserStore = store;
    store.screenReaderSupport.attach();
    store.ui.attach();
    store.browserLocation.attach(fetchGraphql);

    new RouterListener().attach({
      router: Router,
      siteAnalytics: store.siteAnalytics,
      screenReaderSupport: store.screenReaderSupport,
    });
  }

  return store;
}
