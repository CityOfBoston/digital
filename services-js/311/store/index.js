// @flow
/* global window */
/* eslint no-underscore-dangle: 0 */

import { createStore, compose, applyMiddleware } from 'redux';
import type { Store as ReduxStore } from 'redux';
import { middleware as reduxPackMiddleware } from 'redux-pack';
import thunk from 'redux-thunk';
import inject from 'redux-inject';

import type { RequestAdditions } from '../server/next-handlers';
import { reducer } from './modules';
import type { AppState, AppAction } from './modules';
import makeLoopbackGraphql from './deps/loopback-graphql';
import type { LoopbackGraphql } from './deps/loopback-graphql';

// Re-export these to keep things centralized in the “store” module.
export type State = AppState;
export type Action = AppAction;

export type Store = ReduxStore<State, Action>;

export type Deps = {|
  loopbackGraphql: LoopbackGraphql,
|}

// We implement our own Dispatch type to account for the actions and return
// values provided by our middleware. (The Redux built-in type is from Action
// (which must have a type property) to Action. We need to take Actions that
// don't have properties and return pretty much anything, since that’s what
// thunks can return.
export type Dispatch = (action: Action) => any;

// For the browser, we keep a singleton store that can be re-used across pages.
// A module global is the only mechanism for sharing between Next.js pages.
let browserStore = null;

// Factored out to allow for injecting test dependencies and state while
// maintaining the same middleware.
export function makeStore(deps: ?Deps = null, initialState: ?State = null): Store {
  // Support for Redux Devtools Extension: https://github.com/zalmoxisus/redux-devtools-extension
  const composeEnhancers = (process.browser && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

  const enhancers = composeEnhancers(
    applyMiddleware(
      inject(deps),
      thunk,
      reduxPackMiddleware,
    ),
  );

  if (initialState) {
    return createStore(reducer, initialState, enhancers);
  } else {
    return createStore(reducer, enhancers);
  }
}

/**
 * Returns a Redux store for this application. For the server, always returns a
 * fresh store with the given initialState. On the browser, will always return
 * the same store and initialState is ignored after the first call.
 *
 * req is the Node request (expected to contain injected dependencies for server
 * data fetching) that should be provided when creating the store for the
 * server’s getInitialProps calls. Can be null when server rendering or on the
 * client.
 */
export default function getStore(req: ?RequestAdditions, initialState: ?State = null): Store {
  if (process.browser && browserStore) {
    return browserStore;
  }

  const deps = {
    loopbackGraphql: makeLoopbackGraphql(req),
  };

  const store = makeStore(deps, initialState);

  if (process.browser) {
    browserStore = store;
  }

  return store;
}
