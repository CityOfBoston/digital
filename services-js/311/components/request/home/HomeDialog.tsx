import React, { FormEvent } from 'react';
import { action, observable, reaction, runInAction } from 'mobx';
import { observer } from 'mobx-react';
import debounce from 'lodash/debounce';
import Router from 'next/router';

import {
  FetchGraphql,
  GaSiteAnalytics,
  getParam,
} from '@cityofboston/next-client-common';

import { ServiceSummary } from '../../../data/types';
import loadServiceSuggestions from '../../../data/queries/load-service-suggestions';
import loadTopServiceSummaries from '../../../data/queries/load-top-service-summaries';

import FormDialog from '../../common/FormDialog';
import LoadingIcons from '../../common/LoadingIcons';
import { CENTERED_DIALOG_STYLE } from '../../style-constants';
import TranslateDialog from '../translate/TranslateDialog';

import HomePane from './HomePane';
import ChooseServicePane from './ChooseServicePane';
import { GetInitialProps } from '../../../pages/_app';
import { LanguagePreference } from '../../../data/store/BrowserLanguage';
import Ui from '../../../data/store/Ui';
import LiveAgent from '../../../data/store/LiveAgent';

type Stage = 'home' | 'choose';

export type InitialProps = {
  topServiceSummaries: ServiceSummary[];
  description: string;
  stage: Stage;
  bypassTranslateDialog: boolean;
};

export type Props = InitialProps & {
  fetchGraphql: FetchGraphql;
  siteAnalytics: GaSiteAnalytics;
  ui: Ui;
  liveAgent: LiveAgent;
  languages: LanguagePreference[];
};

@observer
export default class HomeDialog extends React.Component<Props> {
  @observable private description: string = '';

  @observable.shallow
  private suggestedServiceSummaries: (ServiceSummary[]) | null = null;
  private lastSuggestedServiceSummariesPromise: Promise<
    ServiceSummary[]
  > | null = null;
  private serviceSuggestionsDisposer: Function | null = null;

  // Called by RequestLayout
  static getInitialProps: GetInitialProps<
    InitialProps,
    'query',
    'fetchGraphql'
  > = async ({ query }, { fetchGraphql }): Promise<InitialProps> => {
    const { stage, description, translate } = query;

    return {
      topServiceSummaries: await loadTopServiceSummaries(fetchGraphql, 5),
      stage: stage === 'choose' ? stage : 'home',
      description: getParam(description, ''),
      bypassTranslateDialog: translate === '0',
    };
  };

  @action
  componentWillMount() {
    this.description = this.props.description;
  }

  componentDidMount() {
    this.serviceSuggestionsDisposer = reaction(
      () => this.description,
      (description: string) => {
        if (description === '') {
          this.suggestedServiceSummaries = [];
        } else {
          this.suggestedServiceSummaries = null;
          this.suggestServices(description);
        }
      },
      {
        fireImmediately: true,
        name: 'update service suggestions from description',
      }
    );

    LoadingIcons.preload();
  }

  componentWillUnmount() {
    if (this.serviceSuggestionsDisposer) {
      this.serviceSuggestionsDisposer();
    }
  }

  suggestServices = debounce(async (description: string) => {
    try {
      this.lastSuggestedServiceSummariesPromise = loadServiceSuggestions(
        this.props.fetchGraphql,
        description
      );

      const suggestedServiceSummaries = await this
        .lastSuggestedServiceSummariesPromise;

      runInAction('suggestServices result', () => {
        this.suggestedServiceSummaries = suggestedServiceSummaries;
      });
    } catch (e) {
      runInAction('suggestServices error', () => {
        this.suggestedServiceSummaries = [];
      });
      throw e;
    }
  }, 500);

  @action.bound
  handleDescriptionChanged(ev: FormEvent<HTMLTextAreaElement>) {
    this.description = ev.currentTarget.value;
  }

  @action.bound
  routeToChoose() {
    const { lastSuggestedServiceSummariesPromise, description } = this;
    const { siteAnalytics } = this.props;

    // We send site analytics only when routing to the choose dialog so that we
    // don't send events for the autocomplete preloading.
    if (lastSuggestedServiceSummariesPromise) {
      lastSuggestedServiceSummariesPromise.then(arr => {
        if (arr.length) {
          siteAnalytics.sendEvent('results', {
            category: 'Prediction',
            label: description,
            value: arr.length,
          });
        } else {
          siteAnalytics.sendEvent('no-results', {
            category: 'Prediction',
            label: description,
            value: 0,
          });
        }
      });
    }

    Router.push(
      `/request?stage=choose&description=${encodeURIComponent(
        this.description.trim()
      )}`,
      `/request?description=${encodeURIComponent(this.description.trim())}`
    ).then(() => window.scrollTo(0, 0));
  }

  render() {
    const { stage, bypassTranslateDialog, languages } = this.props;

    const translateLanguage = TranslateDialog.findLanguage(languages);
    const showTranslate =
      translateLanguage &&
      translateLanguage !== 'en' &&
      !bypassTranslateDialog &&
      stage === 'home';

    const narrow = stage === 'choose';

    if (showTranslate) {
      return <TranslateDialog languages={languages} showContinueInEnglish />;
    }

    return (
      <div className={CENTERED_DIALOG_STYLE}>
        <FormDialog narrow={narrow} noPadding>
          {stage === 'home' && this.renderHome()}
          {stage === 'choose' && this.renderServicePicker()}
        </FormDialog>
      </div>
    );
  }

  renderHome() {
    const { topServiceSummaries, liveAgent, ui } = this.props;
    return (
      <HomePane
        liveAgent={liveAgent}
        ui={ui}
        description={this.description}
        handleDescriptionChanged={this.handleDescriptionChanged}
        nextFn={this.routeToChoose}
        topServiceSummaries={topServiceSummaries}
      />
    );
  }

  renderServicePicker() {
    const { ui, siteAnalytics } = this.props;
    return (
      <ChooseServicePane
        description={this.description}
        suggestedServiceSummaries={this.suggestedServiceSummaries}
        ui={ui}
        siteAnalytics={siteAnalytics}
      />
    );
  }
}
