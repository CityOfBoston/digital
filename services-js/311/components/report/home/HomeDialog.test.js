// @flow

import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import { makeServerContext } from '../../../lib/test/make-context';
import { AppStore } from '../../../data/store';
import HomeDialog from './HomeDialog';
import type { InitialProps } from './HomeDialog';

jest.mock('../../../data/dao/load-top-service-summaries');
jest.mock('../../../data/dao/load-service-suggestions');

const loadTopServiceSummaries: JestMockFn = (require('../../../data/dao/load-top-service-summaries'): any).default;
const loadServiceSuggestions: JestMockFn = (require('../../../data/dao/load-service-suggestions'): any).default;

jest.mock('lodash/debounce');
const debounce: JestMockFn = (require('lodash/debounce'): any);

const SERVICE_SUMMARIES = [{
  name: 'Cosmic Incursion',
  code: 'CSMCINC',
  description: 'Something is threatening the fabric of the universe',
  hasMetadata: true,
  locationRequired: true,
}];

beforeEach(() => {
  // mock implementation always fires, but after a delay
  debounce.mockImplementation((fn) => (...args) => Promise.resolve().then(fn.bind(null, ...args)));
});

describe('home page', () => {
  let initialProps: InitialProps;
  let store;
  let loopbackGraphql;

  beforeEach(async () => {
    store = new AppStore();
    loopbackGraphql = jest.fn();

    loadTopServiceSummaries.mockReturnValue(SERVICE_SUMMARIES);

    const ctx = makeServerContext('/report');
    initialProps = await HomeDialog.getInitialProps(ctx);
  });

  test('getInitialProps', () => {
    expect(initialProps.stage).toEqual('home');
    expect(initialProps.topServiceSummaries).toHaveLength(1);
  });

  test('rendering', () => {
    const component = renderer.create(
      <HomeDialog store={store} loopbackGraphql={loopbackGraphql} {...initialProps} />,
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});

describe('choose page', () => {
  let initialProps: InitialProps;
  let store;
  let loopbackGraphql;

  beforeEach(async () => {
    store = new AppStore();
    loopbackGraphql = jest.fn();

    loadTopServiceSummaries.mockReturnValue(SERVICE_SUMMARIES);

    const ctx = makeServerContext('/report', { stage: 'choose', description: 'Thanos is attacking' });
    initialProps = await HomeDialog.getInitialProps(ctx);
  });

  test('getInitialProps', () => {
    expect(initialProps.stage).toEqual('choose');
    expect(initialProps.description).toEqual('Thanos is attacking');
  });

  test('rendering', () => {
    const component = renderer.create(
      <HomeDialog store={store} loopbackGraphql={loopbackGraphql} {...initialProps} />,
      { createNodeMock: () => document.createElement('div') },
    );
    expect(component.toJSON()).toMatchSnapshot();
  });
});


describe('integration', () => {
  let wrapper;
  let store;

  let resolveSuggestions;

  beforeEach(() => {
    store = new AppStore();

    loadServiceSuggestions.mockReturnValue(new Promise((resolve) => { resolveSuggestions = resolve; }));

    wrapper = mount(<HomeDialog store={store} description="Thanos is attacking" loopbackGraphql={jest.fn()} stage="choose" topServiceSummaries={[]} bypassTranslateDialog />);
  });

  afterEach(() => {
    wrapper.unmount();
  });

  test('description and suggestions', async () => {
    // get on the other side of the debounced lookup triggered by component will mount
    await Promise.resolve();

    expect(loadServiceSuggestions).toHaveBeenCalledWith(expect.anything(), 'Thanos is attacking');

    await resolveSuggestions(SERVICE_SUMMARIES);

    // The service should now appear in the list
    const serviceLink = wrapper.find('a').findWhere((el) => el.text() === 'Cosmic Incursion');
    expect(serviceLink.length).toEqual(1);
    // We look to the parent, which is a Next Link component.
    expect(serviceLink.first().parent().props().href).toEqual('/report?code=CSMCINC&description=Thanos%20is%20attacking');
  });
});
