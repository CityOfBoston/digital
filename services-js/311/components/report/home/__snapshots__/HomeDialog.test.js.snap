// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`choose page rendering 1`] = `
<div>
  <div
    className={
      Object {
        "data-css-1rld2am": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1d2khtc": "",
        }
      }
    >
      <div
        className=" br br-t400 br--y css-182uick"
        style={
          Object {
            "maxWidth": 800,
          }
        }
      >
        <div>
          <div
            className="p-a300 p-a800--xl"
            style={
              Object {
                "paddingBottom": ".75rem",
              }
            }
          >
            <div
              className="sh sh--y"
            >
              <h2
                className="sh-title"
              >
                BOS:311 — File a Report
              </h2>
            </div>
            <div
              className="m-v500 t--intro"
            >
              “
              Thanos is attacking
              ”
            </div>
          </div>
          <div
            className="b b--g p-a300 p-a800--xl"
          >
            <h3
              className="stp"
            >
              How can we help?
            </h3>
            <div
              className="t--info m-v300"
            >
              Matching your problem to BOS:311 services…
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`home page rendering 1`] = `
<div>
  <div
    className={
      Object {
        "data-css-1rld2am": "",
      }
    }
  >
    <div
      className={
        Object {
          "data-css-1d2khtc": "",
        }
      }
    >
      <div
        className=" br br-t400 br--y css-182uick"
        style={Object {}}
      >
        <div>
          <div
            className="p-a300 p-a800--xl"
          >
            <div
              className="sh sh--y"
            >
              <h2
                className="sh-title"
              >
                BOS:311 — File a Report
              </h2>
            </div>
            <div
              className="t--info m-v300"
            >
              Through BOS:311, you can report non-emergency issues with the City.
            </div>
            <div
              className="g m-t500"
            >
              <div
                className="g--7 css-1vm2aiv"
              >
                <div>
                  <h3
                    className="stp m-v100 css-17cwx2g"
                  >
                    Tell us your problem:
                  </h3>
                  <textarea
                    className={
                      Object {
                        "data-css-14mt46e": "",
                      }
                    }
                    defaultValue=""
                    name="description"
                    onBlur={[Function]}
                    onFocus={[Function]}
                    onInput={[Function]}
                    placeholder=""
                  />
                </div>
                <div
                  className="m-t500"
                  style={
                    Object {
                      "textAlign": "right",
                    }
                  }
                >
                  <button
                    className="btn css-1q8b8td"
                    disabled={true}
                    onClick={[Function]}
                  >
                    Start a Report
                  </button>
                </div>
              </div>
              <div
                className="g--1 css-1lffeal"
              >
                <div
                  className="br br-a150 css-7mg7kt"
                />
                <div
                  className="br br-a150 t--info css-u5arvv"
                >
                  or
                </div>
                <div
                  className="br br-a150 css-7mg7kt"
                />
              </div>
              <div
                className="g--4 css-17cwx2g"
              >
                <div
                  className="t--info"
                >
                  You can also start a report by picking one of these popular services:
                </div>
                <ul
                  className="ul m-v300"
                >
                  <li
                    className="t--info"
                  >
                    <a
                      className="m-v100"
                      href="/report/CSMCINC"
                      onClick={[Function]}
                    >
                      Cosmic Incursion
                    </a>
                  </li>
                </ul>
                <div
                  className="t--info"
                >
                  <a
                    href="/services"
                    onClick={[Function]}
                  >
                    See all services…
                  </a>
                </div>
              </div>
            </div>
          </div>
          <div
            className="b b--g p-a300 p-a800--xl"
            style={
              Object {
                "paddingBottom": 0,
                "paddingTop": 0,
              }
            }
          >
            <div
              className="g"
            >
              <form
                acceptCharset="UTF-8"
                action="/lookup"
                className="sf sf--y sf--md g--7 m-v400"
                method="get"
                onFocus={[Function]}
                onSubmit={[Function]}
              >
                <div
                  className="sf-i"
                >
                  <input
                    className="sf-i-f"
                    name="q"
                    onChange={[Function]}
                    placeholder="Search by case ID or keywords…"
                    type="text"
                    value=""
                  />
                  <button
                    className="sf-i-b"
                  >
                    Search
                  </button>
                </div>
              </form>
              <div
                className="g--1"
              />
              <div
                className="g--4 t--info css-12alag6"
              >
                <a
                  href="/search"
                  onClick={[Function]}
                >
                  Browse public cases
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
